src/TCP.cxx

```diff
@@ -0,0 +1,79 @@
+// MIT License
+//
+// Copyright (c) 2021 Felix Moessbauer
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+///
+/// \author Felix Moessbauer
+///
+
+#include "TCP.h"
+#include "InfluxDBException.h"
+#include <string>
+
+namespace influxdb::transports
+{
+
+    TCP::TCP(const std::string& hostname, int port)
+        : mSocket(mIoService, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0))
+    {
+        boost::asio::ip::tcp::resolver resolver(mIoService);
+        boost::asio::ip::tcp::resolver::query query(boost::asio::ip::tcp::v4(), hostname, std::to_string(port));
+        boost::asio::ip::tcp::resolver::iterator resolverInerator = resolver.resolve(query);
+        mEndpoint = *resolverInerator;
+        reconnect();
+    }
+
+    TCP::~TCP()
+    {
+        mSocket.shutdown(boost::asio::ip::tcp::socket::shutdown_both);
+        mSocket.close();
+    }
+
+    bool TCP::is_connected() const
+    {
+        return mSocket.is_open();
+    }
+
+    void TCP::reconnect()
+    {
+        mSocket.connect(mEndpoint);
+        mSocket.wait(boost::asio::ip::tcp::socket::wait_write);
+    }
+
+    void TCP::send(std::string&& message)
+    {
+        try
+        {
+            message.append("\n");
+            const size_t written = mSocket.write_some(boost::asio::buffer(message, message.size()));
+            if (written != message.size())
+            {
+                mSocket.close();
```

Comment:
I'm not sure about this. Completely closing the connection seems a bit hard to me here. Usually one doesn't expect getting the connection teared down on a transmission error.

Reply:
Good point. Given, that this is TCP, the write very likely failed because we lost the connection. By that, it makes sense to close the socket. However, this remains an implementation detail. Maybe we better just throw the exception and add a note that the user should check `is_connected` on write errors.

In general, the implementation initially was written for high data-rates (>10MB / s) at minimal resource consumption on both ends (sender and QuestDB). That's why we keep the TCP connection open to avoid the handshake and connection tracking overhead. However, this design decision is a bit problematic for micro-service architectures, where services restart and we have to reconnect.